File: 2023/go/day1/day1.go
package day1

import (
	"fmt"
	"strconv"
	"strings"
	"unicode"
)

type day1 struct {
	input string
}

func NewDay1(input string) *day1 {
	return &day1{input: input}
}

func Run(input string) int {
	d1 := NewDay1(input)
	fmt.Println("Day 1 Part 1:", d1.Part1())
	return 0
}

func (d day1) Part1() int {
	// for each line in the input
	// scan the line for numbers and form an array
	// the first number of the array and last number
	// of the array are combined and returned
	sum := 0
	for _, line := range strings.Split(d.input, "\n") {
		digits := []rune{}
		for _, char := range line {
			if unicode.IsDigit(char) {
				digits = append(digits, char)
			}
		}
		rowStr := fmt.Sprintf("%c%c", digits[0], digits[len(digits)-1])
		row, _ := strconv.Atoi(rowStr)
		sum += row
	}

	return sum
}

func (d day1) Part2() int {
	// for each line in the input
	// scan the line for numbers and form an array
	// the first number of the array and last number
	// of the array are combined and returned
	sum := 0
	words := [9]string{"one", "two", "three", "four", "five", "six", "seven", "eight", "nine"}
	for _, line := range strings.Split(d.input, "\n") {
		digits := []rune{}
		word := ""
		for _, char := range line {
			if unicode.IsDigit(char) {
				println("digit", string(char-0))
				digits = append(digits, char-48)
				word = ""
			} else if unicode.IsLetter(char) {
				word += string(char)
				fmt.Println(word)
				for index, w := range words {
					if strings.Contains(word, w) {
						digits = append(digits, rune(index+1))
						word = ""
					}
				}
			}
			fmt.Printf("%+v\n", digits)
		}
		fmt.Printf("DIGITS: %+v\n", digits)
		rowStr := fmt.Sprintf("%c%c", digits[0], digits[len(digits)-1])
		fmt.Printf("LINE: %v\n", rowStr)
		row, _ := strconv.Atoi(rowStr)
		sum += row
	}

	return sum
}

File: 2023/python/aoc2023/day1.py
import re


def main(data: str):
    p1 = part1(data)
    p2 = part2(data)
    print(f"Day 1 Part 1: {p1}, Part 2: {p2}")


def part1(data: str) -> int:
    print("Part 1")
    sum = 0

    lines = [line for line in data.split("\n") if line != ""]
    for line in lines:
        digits: list[str] = []
        for i in range(len(line)):
            if line[i].isdigit():
                digits.append(line[i])
        result = int(f"{digits[0]}{digits[-1]}")
        sum += result
    return sum


def part2(data: str):
    print("Part 2")
    sum = 0

    lines = [line for line in data.split("\n") if line != ""]
    num_words = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]

    for line in lines:
        # All numbers in the line with their positions
        num_map: list[tuple[int, int]] = []

        # Digit numbers
        for i in range(len(line)):
            if line[i].isdigit():
                num_map.append((i, int(line[i])))

        # Word numbers
        for i, num in enumerate(num_words):
            positions = [m.start() for m in re.finditer(num, line)]
            for pos in positions:
                num_map.append((pos, i + 1))
        first_num = (0, 0)
        last_num = (0, 0)
        for i, tup in enumerate(num_map):
            if i == 0:
                first_num = tup
                last_num = tup
                continue
            if tup[0] < first_num[0]:
                first_num = tup
            if tup[0] > last_num[0]:
                last_num = tup
        result = int(f"{first_num[1]}{last_num[1]}")
        print(num_map, result)
        sum += result
    return sum
